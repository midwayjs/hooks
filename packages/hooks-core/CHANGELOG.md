# @midwayjs/hooks-core

## 3.0.0-rc.3

### Patch Changes

- Set correct version for examples

## 3.0.0-rc.2

### Patch Changes

- Build after publish

## 3.0.0-rc.1

### Patch Changes

- 71a3115: Release 3.0.0 rc version
