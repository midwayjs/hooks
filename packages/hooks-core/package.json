{
  "name": "@midwayjs/hooks-core",
  "version": "3.0.0-rc.6",
  "description": "Midway Hooks Core",
  "keywords": [
    "midway",
    "hooks",
    "runtime"
  ],
  "author": "Lxxyx <lxxyxzj@gmail.com>",
  "homepage": "https://github.com/midwayjs/hooks",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git@github.com:midwayjs/hooks.git"
  },
  "main": "dist/midwayjs-hooks-core.cjs.js",
  "module": "dist/midwayjs-hooks-core.esm.js",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "asynchronous-local-storage": "^1.0.2",
    "fn-args": "^5.0.0",
    "lodash": "^4.17.21",
    "proper-url-join": "^2.1.1",
    "reflect-metadata": "^0.1.13",
    "upath": "^2.0.1",
    "zod": "^3.11.6",
    "koa-compose": "^4.1.0",
    "debug": "^4.3.3"
  },
  "devDependencies": {
    "@types/lodash": "4.14.178",
    "@types/proper-url-join": "2.1.1",
    "@types/debug": "4.1.7"
  },
  "files": [
    "dist"
  ]
}
