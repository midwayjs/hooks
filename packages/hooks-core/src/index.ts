export * from './runtime'
export * from './router'
export * from './types'
export * from './common'

export * from './api'
export * from './adapter'

export { default as urlJoin } from 'proper-url-join'
