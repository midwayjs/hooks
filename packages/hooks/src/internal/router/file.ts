import kebabCase from 'lodash/kebabCase'
import last from 'lodash/last'
import { extname, join, relative, removeExt, toUnix } from 'upath'
import { AbstractRouter, Route, urlJoin } from '@midwayjs/hooks-core'
import { camelCase } from 'lodash'

export enum RouteKeyword {
  INDEX = 'index',
  CATCH_ALL = '/*',
  DYNAMIC = ':',
}

export interface FileSystemRouterConfig {
  source: string
  routes: Route[]
  legacy?: boolean
}

type Part = {
  content: string
  dynamic: boolean
  qualifier?: string
  catchAll?: boolean
}

export class FileSystemRouter extends AbstractRouter {
  constructor(public config: FileSystemRouterConfig) {
    super()
  }

  getApiDirectory(baseDir: string) {
    return join(this.config.source, baseDir)
  }

  getFunctionId(file: string, functionName: string, isExportDefault: boolean) {
    const relativePath = relative(this.config.source, file)
    // src/apis/lambda/index.ts -> apis-lambda-index
    const id = kebabCase(removeExt(relativePath, extname(relativePath)))
    const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
    return this.config.legacy ? name.toLowerCase() : camelCase(name)
  }

  getRoute(file: string) {
    return this.config.routes.find((route) => {
      const apiDir = this.getApiDirectory(route.baseDir)
      return this.isPathInside(toUnix(file), toUnix(apiDir))
    })
  }

  isApiFile({ file }) {
    const route = this.getRoute(file)
    return !!route
  }

  functionToHttpPath(
    file: string,
    functionName: string,
    exportDefault: boolean
  ) {
    const { baseDir } = this.getRoute(file)

    const filePath = removeExt(
      relative(this.getApiDirectory(baseDir), file),
      extname(file)
    )

    return urlJoin(this.buildUrl(filePath, exportDefault ? '' : functionName), {
      trailingSlash: false,
    })
  }

  buildUrl(file: string, functionName = '') {
    const parts: Part[] = file
      .split(/\[(.+?\(.+?\)|.+?)\]/)
      .map((str, i) => {
        if (!str) return null
        const dynamic = i % 2 === 1

        const [, content, qualifier] = dynamic
          ? /([^(]+)(\(.+\))?$/.exec(str)
          : [null, str, null]

        return {
          content,
          dynamic,
          catchAll: /^\.{3}.+$/.test(content),
          qualifier,
        }
      })
      .filter(Boolean)

    const catchAllIndex = parts.findIndex((part) => part.catchAll)
    if (catchAllIndex !== -1 && catchAllIndex !== parts.length - 1) {
      throw new Error(
        `Catch all routes must be the last part of the path. Current input: ${file}`
      )
    }

    const segments: string[] = []

    for (let [idx, { content, dynamic, catchAll }] of parts.entries()) {
      if (!content) continue

      if (catchAll) {
        if (content.slice(3) !== RouteKeyword.INDEX) {
          segments.push(content.slice(3))
        }
        continue
      }

      /**
       * /[api]/id -> /:api/id
       */
      if (dynamic) {
        segments.push(`${RouteKeyword.DYNAMIC}${content}`)
        continue
      }

      if (idx === parts.length - 1) {
        const contents = content.split('/')
        if (last(contents) === RouteKeyword.INDEX) {
          contents.pop()
          segments.push(contents.join('/'))
          continue
        }
      }

      segments.push(content)
    }

    segments.push(functionName)
    if (catchAllIndex !== -1) {
      segments.push(RouteKeyword.CATCH_ALL)
    }

    return urlJoin.apply(null, [...segments, {}])
  }
}
