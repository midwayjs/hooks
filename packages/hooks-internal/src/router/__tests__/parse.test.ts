import { FileSystemRouter } from '../file'
import { parseApiModule } from '../'
import { Api, Delete, OperatorType } from '@midwayjs/hooks-core'

test('FileSystemRouter should load pure function as api', async () => {
  const mod = {
    get() {},
    post(name: string) {},
    del: Api(Delete(), async () => {}),
  }
  const file = '/src/lambda/index.ts'
  const router = new FileSystemRouter({
    source: '/src',
    routes: [
      {
        baseDir: 'lambda',
      },
    ],
  })
  const apis = parseApiModule(mod, file, router)

  for (const api of apis) {
    expect(Reflect.getMetadata(OperatorType.Trigger, api.fn)).toMatchSnapshot()
  }
})
