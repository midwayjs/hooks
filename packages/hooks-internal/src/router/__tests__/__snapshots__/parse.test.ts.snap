// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSystemRouter should load pure function as api 1`] = `
Object {
  "method": "GET",
  "path": "/get",
  "requestClient": Object {
    "client": "@midwayjs/rpc",
    "fetcher": "http",
  },
  "type": "HTTP",
}
`;

exports[`FileSystemRouter should load pure function as api 2`] = `
Object {
  "method": "POST",
  "path": "/post",
  "requestClient": Object {
    "client": "@midwayjs/rpc",
    "fetcher": "http",
  },
  "type": "HTTP",
}
`;

exports[`FileSystemRouter should load pure function as api 3`] = `
Object {
  "method": "DELETE",
  "path": "/del",
  "requestClient": Object {
    "client": "@midwayjs/rpc",
    "fetcher": "http",
  },
  "type": "HTTP",
}
`;
